src/axom/core/tests/core_array.cpp

```diff
@@ -0,0 +1,1103 @@
+  }
+
+  /* Insert into the front of the Array. */
+  for(IndexType i = n_insert_front - 1; i >= 0; i--)
+  {
+    capacity = calc_new_capacity(v, 1);
+    typename axom::Array<T>::ArrayIterator ret3 = v.insert(v.begin(), 1, i);
+    EXPECT_EQ(ret3, v.begin());
+    size++;
+  }
+
+  /* Check that the insertion worked as expected */
+  EXPECT_EQ(v.capacity(), capacity);
+  EXPECT_EQ(v.size(), size);
+  for(IndexType i = 0; i < size; ++i)
+  {
+    EXPECT_EQ(v[i], i);
+  }
+
+  deallocate(values);
+  values = nullptr;
+}
+
+/*!
+ * \brief Check that emplace() into an Array is working properly.
+ * \param [in] v the Array to check.
+ */
+template <typename T>
+void check_emplace(Array<T>& v)
+{
+  /* Resize the array up to the capacity */
+  IndexType capacity = v.capacity();
+  v.resize(capacity);
+  IndexType size = capacity;
+
+  EXPECT_EQ(v.size(), v.capacity());
+
+  /* Set the existing data in v */
+  for(IndexType i = 0; i < size; ++i)
+  {
+    v[i] = i - 5 * i + 7;
+  }
+
+  /* Emplace a new element, should resize. */
+  IndexType old_capacity = capacity;
+  capacity = calc_new_capacity(v, 1);
+  typename axom::Array<T>::ArrayIterator ret =
+    v.emplace(v.end(), size - 5 * size + 7);
+  size++;
+
+  /* Check that it resized properly */
+  EXPECT_GT(capacity, old_capacity);
+  EXPECT_EQ(v.size(), size);
+  EXPECT_EQ(ret, v.end() - 1);
+  for(IndexType i = 0; i < size; ++i)
+  {
+    EXPECT_EQ(v[i], i - 5 * i + 7);
+  }
+
+  /* Emplace_back 1000 elements */
+  const IndexType n_elements = 1000;
+  for(IndexType i = 0; i < n_elements; ++i)
+  {
+    IndexType i_real = i + size;
+    v.emplace_back(i_real - 5 * i_real + 7);
+  }
+
+  size += n_elements;
+
+  /* Check that it resizes properly */
+  EXPECT_EQ(v.size(), size);
+  for(IndexType i = 0; i < size; ++i)
+  {
+    EXPECT_EQ(v[i], i - 5 * i + 7);
+  }
+
+  capacity = size;
+  v.shrink();
+  IndexType n_insert_front = 100;
+
+  /* Reset the data */
+  T* data_ptr = v.data();
+  for(IndexType i = 0; i < size; ++i)
+  {
+    data_ptr[i] = i + n_insert_front;
+  }
+
+  /* Emplace into the front of the Array. */
+  for(IndexType i = n_insert_front - 1; i >= 0; i--)
+  {
+    capacity = calc_new_capacity(v, 1);
+    typename axom::Array<T>::ArrayIterator ret3 = v.emplace(v.begin(), i);
+    EXPECT_EQ(ret3, v.begin());
+    size++;
+  }
+
+  /* Check that the emplace worked as expected */
+  EXPECT_EQ(v.capacity(), capacity);
+  EXPECT_EQ(v.size(), size);
+  for(IndexType i = 0; i < size; ++i)
+  {
+    EXPECT_EQ(v[i], i);
+  }
+}
+
+template <typename T>
+void check_swap(Array<T>& v)
+{
+  axom::Array<T> v_two(v.size());
+
+  /* Push 0...size elements */
+  for(int i = 0; i < v.size(); i++)
+  {
+    v[i] = i;
+    v_two[i] = -i;
+  }
+
+  /* Create copies */
+  axom::Array<T> v_copy(v);
+  axom::Array<T> v_two_copy(v_two);
+
+  EXPECT_EQ(v, v_copy);
+  EXPECT_EQ(v_two, v_two_copy);
+
+  /* Swap */
+  v.swap(v_two);
+
+  EXPECT_EQ(v_two, v_copy);
+  EXPECT_EQ(v, v_two_copy);
```

Comment:
Perhaps this should instead loop through the array and check that the values have actually been swapped?

Alternatively, it would likely be sufficent to show that
```cpp
EXPECT_NE(v, v_two);
EXPECT_NE(v, v_copy);
```